datafusion/core/src/physical_plan/aggregates/priority_queue.rs

```diff
@@ -0,0 +1,415 @@
+    group_by: PhysicalGroupBy,
+    group_converter: RowConverter,
+    value_converter: RowConverter,
+    priority_map: PriorityQueueBiMap,
+}
+
+impl GroupedPriorityQueueAggregateStream {
+    pub fn new(
+        agg: &AggregateExec,
+        context: Arc<TaskContext>,
+        partition: usize,
+        limit: usize,
+    ) -> Result<Self> {
+        let agg_schema = Arc::clone(&agg.schema);
+        let agg_group_by = agg.group_by.clone();
+
+        let input = agg.input.execute(partition, Arc::clone(&context))?;
+
+        let aggregate_arguments =
+            aggregate_expressions(&agg.aggr_expr, &agg.mode, agg_group_by.expr.len())?;
+
+        let (field, descending) = agg
+            .get_minmax_desc()
+            .ok_or_else(|| DataFusionError::Execution("Min/max required".to_string()))?;
+        let sort = SortOptions {
+            descending,
+            nulls_first: false,
+        };
+        let value_fields =
+            vec![SortField::new_with_options(field.data_type().clone(), sort)];
+        let value_converter = RowConverter::new(value_fields.clone())?;
+
+        // TODO: filters
+
+        let group_schema = group_schema(&agg_schema, agg_group_by.expr.len());
+        let group_fields = group_schema
+            .fields()
+            .iter()
+            .map(|f| SortField::new(f.data_type().clone()))
+            .collect::<Vec<_>>();
+        let group_converter = RowConverter::new(group_fields.clone())?;
+
+        Ok(GroupedPriorityQueueAggregateStream {
+            schema: agg_schema,
+            input,
+            aggregate_arguments,
+            group_by: agg_group_by,
+            priority_map: PriorityQueueBiMap::new(limit),
+            group_converter,
+            value_converter,
+        })
+    }
+}
+
+impl RecordBatchStream for GroupedPriorityQueueAggregateStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+impl Stream for GroupedPriorityQueueAggregateStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        while let Poll::Ready(res) = self.input.poll_next_unpin(cx) {
+            match res {
+                // got a batch, convert to rows and append to our TreeMap
+                Some(Ok(batch)) => {
+                    let group_by_values = evaluate_group_by(&self.group_by, &batch)?;
+                    let group_by_values = match group_by_values.as_slice() {
+                        [] => {
+                            Err(DataFusionError::Execution("vals required".to_string()))?
+                        }
+                        [vals] => vals,
+                        _ => {
+                            Err(DataFusionError::Execution("1 val required".to_string()))?
+                        }
+                    };
+                    let input_values = evaluate_many(&self.aggregate_arguments, &batch)?;
```

Comment:
I guess it would be most optimal to:
1. execute the min/max aggrgate first
2. create a mask based on resp. the max/min value (i.e. do they make it to the top `n` values for each group)
3. only do row conversion and rest of the logic on filtered values